File: 2014-Oxygenium/Java/Solution.java
class Solution {
    class Mem {
        int value;
        int posA;
        Mem(int value, int posA) {
            this.value = value;
            this.posA = posA;
        }
    }

    interface Cmp {
        boolean compare (Comparable a, Comparable b);
    }

    class GreaterEqual implements Cmp {
        public boolean compare( Comparable a, Comparable b) {
            int r = a.compareTo(b);
            if( (r == 1) || (r == 0) ) return true;
            return false;
        };
    }

    class LessEqual implements Cmp {
        public boolean compare( Comparable a, Comparable b) {
            int r = a.compareTo(b);
            if( (r == 0) || (r == -1) ) return true;
            return false;
        };
    }

    public < T extends Cmp > int add_elem(T cmp, Mem a, Mem mem[], int begin, int end)
    {
        while( (end >= begin) && cmp.compare( mem[end].value, a.value) ) {
            end -= 1;
        }

        end += 1;
        mem[end] = a;
        return end;
    }


    public int solution(int K, int[] A) {
        final int max_int = 1_000_000_000;
        int N = A.length;

        Mem[] mini = new Mem[N+1];
        Mem[] maxi = new Mem[N+1];
        int mini_first = 0, mini_last = -1;
        int maxi_first = 0, maxi_last = -1;
        int j = 0;
        int result = 0;
        GreaterEqual ge = new GreaterEqual();
        LessEqual lt = new LessEqual();

        for( int i = 0; i < N; ++i ) {
            while( j < N ) {
                mini_last = add_elem(ge, new Mem(A[j], j), mini, mini_first, mini_last );
                maxi_last = add_elem(lt, new Mem(A[j], j), maxi, maxi_first, maxi_last );

                if( (maxi[maxi_first].value - mini[mini_first].value) <= K ) {
                    j += 1;
                }
                else {
                    break;
                }
            }

            result += (j - i);
            if(result >= max_int) return max_int;
            if(mini[mini_first].posA == i) mini_first += 1;
            if(maxi[maxi_first].posA == i) maxi_first += 1;
        }

        return result;
    }
}

File: 2014-Oxygenium/C/solution.c
#include <stdio.h>
#include <stdlib.h>

typedef struct _VP {
	int val;
	int pos;
} VP;

typedef int cmp_ft(int, int);

int greater_equal(int a, int b)
{
	return a >= b;
}

int less_equal(int a, int b)
{
	return a <= b;
}

int add_element( VP vp, VP* arr, int begin, int end, cmp_ft cmp )
{
	while( (end >= begin) && cmp(arr[end].val, vp.val) ) {
		end--;
	}
	end++;
	arr[end] = vp;
	return end;
}

int solution( int K, int A[], int N )
{
	int result = 0;
	int i = 0, j = 0;

	VP* mini = malloc(N * sizeof(VP));
	VP* maxi = malloc(N * sizeof(VP));

	int mini_begin = 0, mini_end = -1;
	int maxi_begin = 0, maxi_end = -1;
	VP vp;

	for( i = 0; i < N; ++i ) {
		while( j < N ) {
			vp.val = A[j];
			vp.pos = j;

			mini_end = add_element( vp, mini, mini_begin, mini_end, greater_equal );
			maxi_end = add_element( vp, maxi, maxi_begin, maxi_end, less_equal );
			if( (maxi[maxi_begin].val - mini[mini_begin].val) <= K ) {
				j++;
			}
			else {
				break;
			}
		}
		result += (j -i);
		if( result > 1e9) return 1e9;

		if( mini[mini_begin].pos == i ) { mini_begin++; };
		if( maxi[maxi_begin].pos == i ) { maxi_begin++; };
	}

	return result;
}

#ifdef MY_COMPUTER
int main() {
	int A[] = {3, 5, 7, 6, 3};
	int K = 2;
	printf( "%d\n", solution( K, A, sizeof(A)/sizeof(int) ));
	return 0;
}
#endif
